src/typechecker.erl

```diff
@@ -1119,6 +1119,71 @@ expect_tuple_union([], AccTy, AccCs, any, N) ->
 expect_tuple_union([], AccTy, AccCs, _NoAny, _N) ->
     {AccTy, AccCs}.
 
+purge_binary(ResTy) ->
+    case ResTy of
+        {type, _, binary, [{integer, _, 0}, {integer, _, _N}]} ->
+            %% The result is a multiple of N bits.
+            %% Expr must be a multiple of N bits too.
+            ResTy;
+        {type, _, binary, [{integer, _, M}, {integer, _, _N}]}
+          when M > 0 ->
+            %% The result is a binary with a minimum size of M. This
+            %% requires that the generators are non-empty. We don't
+            %% check that. At least, we can check that Gen is a
+            %% bitstring.
+            %% TODO: properly check these values
+            {type, erl_anno:new(0), binary,
+             [{integer, erl_anno:new(0), 0},
+              {integer, erl_anno:new(0), 1}]}
+    end.
+
+expect_binary_type(?type(any), _) ->
+    any;
+expect_binary_type(ElemTy = {type, _, binary, _}, _) ->
+    {elem_ty, ElemTy, constraints:empty()};
+expect_binary_type(Union = {type, _, union, UnionTys}, TEnv) ->
+    {Tys, Cs} = expect_binary_union(UnionTys, [], constraints:empty(), TEnv),
+    case Tys of
+        [] ->
+            {type_error, Union};
+        [Ty] ->
+            {elem_ty, Ty, Cs};
+        _ ->
+            {elem_tys, Tys, Cs}
+    end;
+expect_binary_type({var, _, Var}, _) ->
+    TyVar = new_type_var(),
+    {elem_ty
+    ,{var, erl_anno:new(0), TyVar}
+    ,constraints:add_var(TyVar,
+      constraints:upper(Var, {type, erl_anno:new(0), binary, [TyVar]}))
+    };
+expect_binary_type(Ty, _) ->
+    {type_error, Ty}.
+
+expect_binary_union([Ty|Tys], AccTy, AccCs, TEnv) ->
+    case expect_binary_type(normalize(Ty, TEnv), TEnv) of
+        {type_error, _} ->
+            expect_binary_union(Tys, AccTy, AccCs, TEnv);
+        any ->
+            expect_binary_union(Tys
+                                ,[type(any) | AccTy]
+                                ,AccCs
+                                ,TEnv);
+        {elem_ty, NTy, Cs} ->
```

Comment:
My memory is rather foggy about how this works but from what I recall, handling `any()` in this way doesn't actually work. Look at how it is implemented for lists and tuples. There we always return just `any()` as the element type if in one of the unions there is a case that the element type is `any()`.

Reply:
@josefs I don't know about the case for lists, but we have working examples of handling unions with any: https://github.com/josefs/Gradualizer/blob/master/test/should_fail/union_with_any.erl

Reply:
I think that in this time I did it pretty much as in the case for lists: https://github.com/josefs/Gradualizer/blob/0b4c6b0b78741bb77f1226db7f316bdf1b38787f/src/typechecker.erl#L1014 `any` doesn't make the rest of the list be ignored, right?

Reply:
Yes, you're right. I read the code wrong.